src/main/java/org/elasticsearch/action/admin/indices/seal/TransportSealIndicesAction.java

```diff
@@ -0,0 +1,168 @@
+        super(settings, SealIndicesAction.NAME, threadPool, transportService, actionFilters, SealIndicesRequest.class);
+        this.syncedFlushService = syncedFlushService;
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+        transportService.registerRequestHandler(SealIndicesAction.NAME, ShardSyncedFlushRequest.class, ThreadPool.Names.GENERIC, new ShardSyncedFlushHandler());
+    }
+
+
+    @Override
+    protected void doExecute(final SealIndicesRequest request, final ActionListener<SealIndicesResponse> listener) {
+        // we need to execute in a thread here because getSyncedFlushResults(request, state) blocks
+        ClusterState state = clusterService.state();
+        String[] concreteIndices = state.metaData().concreteIndices(request.indicesOptions(), request.indices());
+        GroupShardsIterator primaries = state.routingTable().activePrimaryShardsGrouped(concreteIndices, false);
+        final CountDown countDown = new CountDown(primaries.size());
+        DiscoveryNode localNode = state.getNodes().localNode();
+        final Set<SyncedFlushService.SyncedFlushResult> results = ConcurrentCollections.newConcurrentSet();
+        for (ShardIterator shard : primaries) {
+            final ShardId shardId = shard.shardId();
+            transportService.sendRequest(localNode, SealIndicesAction.NAME, new ShardSyncedFlushRequest(shardId),
+                    new BaseTransportResponseHandler<SyncedFlushService.SyncedFlushResult>() {
+                        @Override
+                        public SyncedFlushService.SyncedFlushResult newInstance() {
+                            return new SyncedFlushService.SyncedFlushResult();
+                        }
+
+                        @Override
+                        public void handleResponse(SyncedFlushService.SyncedFlushResult response) {
+                            results.add(response);
+                            if (countDown.countDown()) {
+                                listener.onResponse(new SealIndicesResponse(results));
+                            }
+                        }
+
+                        @Override
+                        public void handleException(TransportException exp) {
+                            logger.debug("{} unexpected error while executing synced flush", shardId);
+                            results.add(new SyncedFlushService.SyncedFlushResult(shardId, exp.getMessage()));
+                            if (countDown.countDown()) {
+                                listener.onResponse(new SealIndicesResponse(results));
+                            }
+                        }
+
+                        @Override
+                        public String executor() {
+                            return ThreadPool.Names.GENERIC;
```

Comment:
I think this can be `SAME` in this case as the stuff we are executing in the handler is super cheap?